### ISSUE TITLE
Giraffe bug about faster_cap: `!isinf(c.back())' failed.

### ISSUE THREAD

**Issue by jmonlong** [vgteam/vg#3003]
I was mapping reads real short reads and vg giraffe stopped with an error.

My command: `vg giraffe -x hg38-hsvlr_srdedup17_aug.xg -dhg38-hsvlr_srdedup17_aug.dist -m hg38-hsvlr_srdedup17_aug.k29.w11.N32.min --gbwt-name hg38-hsvlr_srdedup17_aug.N32.gbwt -f reads.R1.fastq.gz -f reads.R2.fastq.gz -p -t 16 -b fast -o gaf -C 500 > out.gaf`

The log and error message:

```
Initializing MinimizerMapper
Loading and initialization: 322.335 seconds
Mapping reads to "-" (GAF)
--hit-cap 10
--hard-hit-cap 500
--score-fraction 0.5
--max-extensions 400
--max-alignments 8
--cluster-score 50
--pad-cluster-score 0
--cluster-coverage 0.2
--extension-score 1
--extension-set 20
--max-multimaps 1
--distance-limit 200
--paired-distance-limit 2
--rescue-subgraph-size 4
--rescue-attempts 15
--rescue-algorithm dozeu
ERROR: Signal 7 occurred. VG has crashed. Run 'vg bugs --new' to report a bug.
Stack trace path: /tmp/vg_crash_p2HMoj/stacktrace.txt
Please include the stack trace file in your bug report!
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
```

This is using vg commit 43e408c (quay.io/vgteam/vg:v1.27.0-25-g43e408ca1-t361-run)

The indexes are in s3://vg-k8s/users/jmonlong/hsvlr_srdedup17_aug/indexes_release/

I'm in the process of reproducing it to extract the stacktrace.

**Comment by jmonlong** [comment]
The stack trace is 

```
Crash report for vg v1.27.0 "Deliceto"Stack trace (most recent call last) in thread 743:
#14   Object "", at 0xffffffffffffffff, in
#13   Object "/home/jupyter-user/notebooks/vg", at 0x1d3802e, in __clone
#12   Object "/home/jupyter-user/notebooks/vg", at 0x1259c0a, in start_thread
#11   Object "/home/jupyter-user/notebooks/vg", at 0x1c92745, in gomp_thread_start
#10   Object "/home/jupyter-user/notebooks/vg", at 0x101b192, in vg::io::paired_for_each_parallel_after_wait(std::function<b
ool (vg::Alignment&, vg::Alignment&)>, std::function<void (vg::Alignment&, vg::Alignment&)>, std::function<bool ()>, unsigne
d long) [clone ._omp_fn.2]
#9    Object "/home/jupyter-user/notebooks/vg", at 0x1c94bab, in gomp_team_barrier_wait_end
#8    Object "/home/jupyter-user/notebooks/vg", at 0x1c8c119, in gomp_barrier_handle_tasks
#7    Object "/home/jupyter-user/notebooks/vg", at 0x101ade8, in vg::io::paired_for_each_parallel_after_wait(std::function<b
ool (vg::Alignment&, vg::Alignment&)>, std::function<void (vg::Alignment&, vg::Alignment&)>, std::function<bool ()>, unsigne
d long) [clone ._omp_fn.3]
#6    Object "/home/jupyter-user/notebooks/vg", at 0xa65d90, in std::_Function_handler<void (vg::Alignment&, vg::Alignment&)
, main_giraffe(int, char**)::{lambda()#1}::operator()() const::{lambda(vg::Alignment&, vg::Alignment&)#2}>::_M_invoke(std::_
Any_data const&, vg::Alignment&, vg::Alignment&)#5    Object "/home/jupyter-user/notebooks/vg", at 0xb256d0, in vg::MinimizerMapper::map_paired(vg::Alignment&, vg::Alignmen
t&, std::vector<std::pair<vg::Alignment, vg::Alignment>, std::allocator<std::pair<vg::Alignment, vg::Alignment> > >&)
#4    Object "/home/jupyter-user/notebooks/vg", at 0xb21464, in vg::MinimizerMapper::map_paired(vg::Alignment&, vg::Alignmen
t&)
#3    Object "/home/jupyter-user/notebooks/vg", at 0xb26f07, in vg::MinimizerMapper::map_paired(vg::Alignment&, vg::Alignment&)::{lambda(unsigned long)#13}::operator()(unsigned long) const
#2    Object "/home/jupyter-user/notebooks/vg", at 0xb25ded, in vg::MinimizerMapper::attempt_rescue(vg::Alignment const&, vg::Alignment&, std::vector<vg::MinimizerMapper::Minimizer, std::allocator<vg::MinimizerMapper::Minimizer> > const&, bool)
#1    Object "/home/jupyter-user/notebooks/vg", at 0xdf7a32, in vg::MinimumDistanceIndex::subgraph_in_range(vg::Path const&, handlegraph::HandleGraph const*, long, long, std::unordered_set<long long, std::hash<long long>, std::equal_to<long long>,std::allocator<long long> >&, bool)
#0    Object "/home/jupyter-user/notebooks/vg", at 0x12f6da8, in gbwtgraph::CachedGBWTGraph::get_length(handlegraph::handle_t const&) const
```

**Comment by xchang1** [comment]
The stack trace looks like a different error to me. @adamnovak what do you think? The original error looks like a problem with the mapq cap, the stack trace looks like a problem getting the rescue subgraph

**Comment by jmonlong** [comment]
I'm narrowing down to a subset of problematic reads. It doesn't always throw the Assertion failed message, or different numbers sometimes, but the stack trace looks similar all the time I think (if that makes any sense to you?). For example, here is what I just got:

```
Initializing MinimizerMapper
Loading and initialization: 294.207 seconds
Mapping reads to "-" (GAF)
--hit-cap 10
--hard-hit-cap 500
--score-fraction 0.5
--max-extensions 400
--max-alignments 8
--cluster-score 50
--pad-cluster-score 0
--cluster-coverage 0.2
--extension-score 1
--extension-set 20
--max-multimaps 1
--distance-limit 200
--paired-distance-limit 2
--rescue-subgraph-size 4
--rescue-attempts 15
--rescue-algorithm dozeu
ERROR: Signal 7 occurred. VG has crashed. Run 'vg bugs --new' to report a bug.
Stack trace path: /tmp/vg_crash_AM4KCV/stacktrace.txt
Please include the stack trace file in your bug report!
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
vg: src/minimizer_mapper.cpp:2105: static double vg::MinimizerMapper::faster_cap(const std::vector<vg::MinimizerMapper::Minimizer>&, std::vector<long unsigned int>&, const string&, const string&): Assertion `!isinf(c.back())' failed.
````

With stacktrace: 

```
Crash report for vg v1.27.0 "Deliceto"
Stack trace (most recent call last) in thread 137:
#14   Object "", at 0xffffffffffffffff, in
#13   Object "/home/jupyter-user/notebooks/vg", at 0x1d3802e, in __clone
#12   Object "/home/jupyter-user/notebooks/vg", at 0x1259c0a, in start_thread
#11   Object "/home/jupyter-user/notebooks/vg", at 0x1c92745, in gomp_thread_start
#10   Object "/home/jupyter-user/notebooks/vg", at 0x101b192, in vg::io::paired_for_each_parallel_after_wait(std::function<bool (vg::Alignment&, vg::Alignment&)>, std::function<void (vg::Alignment&, vg::Alignment&)>, std::function<bool ()>, unsigned long) [clone ._omp_fn.2]
#9    Object "/home/jupyter-user/notebooks/vg", at 0x1c94bab, in gomp_team_barrier_wait_end
#8    Object "/home/jupyter-user/notebooks/vg", at 0x1c8c119, in gomp_barrier_handle_tasks
#7    Object "/home/jupyter-user/notebooks/vg", at 0x101ade8, in vg::io::paired_for_each_parallel_after_wait(std::function<bool (vg::Alignment&, vg::Alignment&)>, std::function<void (vg::Alignment&, vg::Alignment&)>, std::function<bool ()>, unsigned long) [clone ._omp_fn.3]
#6    Object "/home/jupyter-user/notebooks/vg", at 0xa65d90, in std::_Function_handler<void (vg::Alignment&, vg::Alignment&), main_giraffe(int, char**)::{lambda()#1}::operator()() const::{lambda(vg::Alignment&, vg::Alignment&)#2}>::_M_invoke(std::_Any_data const&, vg::Alignment&, vg::Alignment&)
#5    Object "/home/jupyter-user/notebooks/vg", at 0xb256d0, in vg::MinimizerMapper::map_paired(vg::Alignment&, vg::Alignment&, std::vector<std::pair<vg::Alignment, vg::Alignment>, std::allocator<std::pair<vg::Alignment, vg::Alignment> > >&)
#4    Object "/home/jupyter-user/notebooks/vg", at 0xb21464, in vg::MinimizerMapper::map_paired(vg::Alignment&, vg::Alignment&)
#3    Object "/home/jupyter-user/notebooks/vg", at 0xb26f07, in vg::MinimizerMapper::map_paired(vg::Alignment&, vg::Alignment&)::{lambda(unsigned long)#13}::operator()(unsigned long) const
#2    Object "/home/jupyter-user/notebooks/vg", at 0xb25ded, in vg::MinimizerMapper::attempt_rescue(vg::Alignment const&, vg::Alignment&, std::vector<vg::MinimizerMapper::Minimizer, std::allocator<vg::MinimizerMapper::Minimizer> > const&, bool)
#1    Object "/home/jupyter-user/notebooks/vg", at 0xdf7a32, in vg::MinimumDistanceIndex::subgraph_in_range(vg::Path const&, handlegraph::HandleGraph const*, long, long, std::unordered_set<long long, std::hash<long long>, std::equal_to<long long>,std::allocator<long long> >&, bool)
#0    Object "/home/jupyter-user/notebooks/vg", at 0x12f6da8, in gbwtgraph::CachedGBWTGraph::get_length(handlegraph::handle_t const&) const
````

**Comment by adamnovak** [comment]
I bet the assert is failing, sending a signal, and getting the signal collected by a different thread, which happens to be collecting a rescue graph, and then prints *its* stack trace.

The assert is failing here:

https://github.com/vgteam/vg/blob/master/src/minimizer_mapper.cpp#L2110

We need to find the offending read pair (maybe by turning the assert into a real error message that prints the sequence so we can grep for it), run it in a Giraffe with `#define debug` so we can see what the DP is doing to get an infinite result, and then fix it.

**Comment by adamnovak** [comment]
Wait that doesn't make any sense. Assertion failing is signal 6 (abort); I think 7 (which the log indicates) is a bus error (which is basically a segfault).

Maybe they're somehow caused by the same read?

**Comment by xchang1** [comment]
This seems to have been fixed by #3006. I think it was trying to get a rescue subgraph from an empty alignment but it shouldn't get an empty alignment anymore

**Comment by jmonlong** [comment]
Just reporting that I've had this error in two samples since then (out of >3K samples) but this time I couldn't reproduce it and **re-running "fixed" the issue**. 

For record keeping or if we want to investigate this, one sample was HG01256 from the 1000GP high depth dataset from NYGC: ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR324/ERR3241866/HG01256.final.cram
